File: Napakalaki_Ruby/lib/specific_bad_consequence.rb
# encoding: utf-8

=begin
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
=end

require_relative 'bad_consequence'

class SpecificBadConsequence < BadConsequence

  def initialize(t, l, v, h)
    super(t, l)
    @specific_visible_treasures = v
    @specific_hidden_treasures = h
  end
  
  public_class_method :new
  
  attr_reader :specific_visible_treasures, :specific_hidden_treasures

  #Devuelve true cuando el mal rollo que tiene que cumplir el jugador está vacío, eso
  #significa que el conjunto de atributos del mal rollo indican que no hay mal rollo que
  #cumplir
  def is_empty
    empty = false    
    empty = true if (@specific_hidden_treasures.empty? && @specific_visible_treasures.empty?)
    return empty
  end

  def substract_visible_treasure(t)
    if(!@specific_visible_treasures.empty?)
      @specific_visible_treasures.delete(t.type)
    end
  end

  def substract_hidden_treasure(t)
    if(!@specific_hidden_treasures.empty?)
      @specific_hidden_treasures.delete(t.type)
    end
  end

  def adjust_to_fit_treasure_lists(v, h)
    t = @text
    l = @levels
    result_specific_visible_treasures = Array.new
    result_specific_hidden_treasures = Array.new

    if(!@specific_visible_treasures.empty? || !@specific_hidden_treasures.empty?)
      @specific_visible_treasures.each do |m_v_treasure|
        if (v.include?(m_v_treasure))
          result_specific_visible_treasures << m_v_treasure
        end
      end
      @specific_hidden_treasures.each do |m_h_treasure|
        if (h.include?(m_h_treasure))
          result_specific_hidden_treasures << m_h_treasure
        end
      end
      
      b_c = SpecificBadConsequence.new(t, l, result_specific_visible_treasures, result_specific_hidden_treasures)
    else
      b_c = SpecificBadConsequence.new(t, l, Array.new(0), Array.new(0))
    end

    return b_c
  end

  def to_s
    "Mal rollo: #{@text} \n Niveles: #{@levels} \n Muerte: #{@death}"
  end
end

File: Napakalaki_Java/src/napakalaki/SpecificBadConsequence.java
/*
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
*/
package napakalaki;

import java.util.ArrayList;

public class SpecificBadConsequence extends BadConsequence {
    //Atributos
    protected ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    protected ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    
    //Métodos
    public SpecificBadConsequence(String t, int l, ArrayList<TreasureKind> v, ArrayList<TreasureKind> h) {
        super(t, l);
        this.specificVisibleTreasures = v;
        this.specificHiddenTreasures = h;
    }
    
    @Override
    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return this.specificVisibleTreasures;
    }
    
    @Override
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return this.specificHiddenTreasures;
    }
    
    //Devuelve true cuando el mal rollo que tiene que cumplir el jugador está vacío, eso
    //significa que el conjunto de atributos del mal rollo indican que no hay mal rollo que
    //cumplir.
    @Override
    public boolean isEmpty() {
        boolean empty = false;
        if (this.specificHiddenTreasures.isEmpty() && this.specificVisibleTreasures.isEmpty())
            empty = true;
        return empty;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t) {
        if(!this.specificVisibleTreasures.isEmpty())
            this.specificVisibleTreasures.remove(t.getType());
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        if(!this.specificHiddenTreasures.isEmpty())
            this.specificHiddenTreasures.remove(t.getType());
    }
    
    @Override
    public BadConsequence adjustToFitTreasureList(ArrayList<Treasure> v, ArrayList<Treasure> h) {
        String t = this.text;
        int l = this.levels;
        ArrayList<TreasureKind> resultSpecificVisibleTreasures = new ArrayList();
        ArrayList<TreasureKind> resultSpecificHiddenTreasures = new ArrayList();
        BadConsequence bC;
        
        if(!this.specificVisibleTreasures.isEmpty() || !this.specificHiddenTreasures.isEmpty()) {
            for(TreasureKind mVTreasure : this.specificVisibleTreasures) {
                if(v.contains(mVTreasure))
                    resultSpecificVisibleTreasures.add(mVTreasure);
            }
            for(TreasureKind mHTreasure : this.specificHiddenTreasures) {
                if(h.contains(mHTreasure))
                    resultSpecificHiddenTreasures.add(mHTreasure);
            }
            bC = new SpecificBadConsequence(t, l, resultSpecificVisibleTreasures, resultSpecificHiddenTreasures);
        }
        else
            bC = new SpecificBadConsequence(t, l, new ArrayList(), new ArrayList());
        
        return bC;
    }
    
    @Override
    public String toString() {
        String malrollo = "Mal rollo específico: \n" + "Nombre: " + this.text + "\n";
        if (this.levels != 0)
            malrollo += "Niveles perdidos: " + this.levels + "\n";
        if(!this.specificVisibleTreasures.isEmpty())
            malrollo += "Tesoros visibles perdidos: " + this.specificVisibleTreasures.toString() + "\n";
        if(!this.specificHiddenTreasures.isEmpty())
            malrollo += "Tesoros ocultos perdidos: " + this.specificHiddenTreasures.toString() + "\n";
    
        return malrollo;
    }
}
